jam/server/api/v2/base.py

```diff
@@ -0,0 +1,296 @@
+        v1.extend([v.plural, r'(?P<{}_id>{})'.format(v.name, ID_RE)])
+
+    relationships = '/v2/{}/{}/(?P<relationship>{})/?'.format(
+        view.plural,
+        ':'.join(v2 + [selector]),
+        '|'.join(serializer.relations.keys())
+    )
+
+    v1.append(view.plural)
+
+    v1 = '/v1/' + '/'.join(v1) + ENDING.format(selector)
+    v2 = '/v2/' + view.plural + ENDING.format(':'.join(v2 + [selector]))
+
+    endpoints = [
+        (v1, ResourceHandler, kwargs),
+        (v2, ResourceHandler, kwargs),
+    ]
+
+    if serializer.relations:
+        endpoints.append((relationships, ResourceHandler, kwargs))
+
+    return endpoints
+
+
+class ResourceHandler(JSONAPIHandler):
+    TYPE_DATA_SET = {
+        'PUT': {},
+        'GET': {type(None)},
+        'DELETE': {type(None)},
+        'POST': {dict, list},
+        'PATCH': {dict, list},
+    }
+
+    def get_current_user(self):
+        try:
+            return User(
+                self.request.headers.get('Authorization') or
+                self.get_query_argument('token', default=None)
+            )
+        except jwt.ExpiredSignatureError:
+            return User(None)
+
+    def serialize(self, inst):
+        return self._serializer.serialize(self.request, inst, *self._view.parents)
+
+    def initialize(self, view, serializer):
+        self._view = None
+        self._view_class = view
+        self._serializer = serializer
+
+    def prepare(self):
+        super().prepare()
+        if self.request.method == 'OPTIONS':
+            return  # Dont do anything for OPTIONS requests
+
+        # if type(self.json) not in self.TYPE_DATA_SET[self.request.method]:
+        #     raise Exception()
+
+        loaded = []
+        try:
+            for view in self._view_class.lineage():
+                key = view.name + '_id'
+                if self.path_kwargs[key] is None:
+                    break
+                loaded.append(view.load(self.path_kwargs[key], *loaded))
+        except exceptions.NotFound as e:
+            err = e
+        else:
+            err = None
+
+        self._view = self._view_class(*loaded)
+
+        if 'relationship' in self.path_kwargs:
+            relationship = self._serializer.relations[self.path_kwargs['relationship']]
+            self._view = relationship.view(*loaded)
+            self._serializer = relationship.serializer()
+
+        permissions = Permissions.get_permissions(self.current_user, *loaded)
+        required_permissions = self._view.get_permissions(self.request)
+
+        # For use later on
+        self.current_user.permissions = permissions
+
+        if (required_permissions & permissions) != required_permissions:
+            if self.current_user.uid is None:
+                raise exceptions.Unauthorized()
+            raise exceptions.Forbidden(required_permissions)
+
+        if err:
+            raise err
+
+        if self.request.method in ('GET', 'DELETE'):
```

Comment:
This seems like it could be moved all the way up with the `'OPTIONS'` check at the top of the prepare.